wdl-engine/src/stdlib/size.rs

```diff
@@ -0,0 +1,300 @@
+        CompoundValue::Array(array) => Ok(array.elements().iter().try_fold(0.0, |t, e| {
+            anyhow::Ok(t + calculate_disk_size(e, unit, cwd)?)
+        })?),
+        CompoundValue::Map(map) => Ok(map.elements().iter().try_fold(0.0, |t, (k, v)| {
+            anyhow::Ok(t + primitive_disk_size(k, unit, cwd)? + calculate_disk_size(v, unit, cwd)?)
+        })?),
+        CompoundValue::Object(object) => {
+            Ok(object.members().iter().try_fold(0.0, |t, (_, v)| {
+                anyhow::Ok(t + calculate_disk_size(v, unit, cwd)?)
+            })?)
+        }
+        CompoundValue::Struct(s) => Ok(s.members().iter().try_fold(0.0, |t, (_, v)| {
+            anyhow::Ok(t + calculate_disk_size(v, unit, cwd)?)
+        })?),
+    }
+}
+
+/// Calculates the size of the given directory in the given unit.
+fn calculate_directory_size(path: &Path, unit: StorageUnit) -> Result<f64> {
+    // Don't follow symlinks as a security measure
+    let metadata = path.symlink_metadata().with_context(|| {
+        format!(
+            "failed to read metadata for directory `{path}`",
+            path = path.display()
+        )
+    })?;
+
+    if !metadata.is_dir() {
+        bail!("path `{path}` is not a directory", path = path.display());
+    }
+
+    // Create a queue for processing directories
+    let mut queue: Vec<Cow<'_, Path>> = Vec::new();
+    queue.push(path.into());
+
+    // Process each directory in the queue, adding the sizes of its files
+    let mut size = 0.0;
+    while let Some(path) = queue.pop() {
+        for entry in fs::read_dir(&path)? {
+            let entry = entry.with_context(|| {
+                format!(
+                    "failed to read entry of directory `{path}`",
+                    path = path.display()
+                )
+            })?;
+
+            // Note: `DirEntry::metadata` doesn't follow symlinks
+            let metadata = entry.metadata().with_context(|| {
+                format!(
+                    "failed to read metadata for file `{path}`",
+                    path = entry.path().display()
+                )
+            })?;
+            if metadata.is_dir() {
+                queue.push(entry.path().into());
+            } else {
+                size += unit.convert(metadata.len());
+            }
+        }
+    }
+
+    Ok(size)
+}
+
+/// Gets the function describing `size`.
+pub const fn descriptor() -> Function {
+    Function::new(
+        const {
+            &[
+                Signature::new("(None, <String>) -> Float", size),
+                Signature::new("(File?, <String>) -> Float", size),
+                Signature::new("(String?, <String>) -> Float", size),
+                Signature::new("(Directory?, <String>) -> Float", size),
+                Signature::new(
+                    "(X, <String>) -> Float where `X`: any compound type that recursively \
+                     contains a `File` or `Directory`",
```

Comment:
So what if not all the members are `File` or `Directory`? Are other primitive types just ignored?

Reply:
That's correct, it only looks at `File` and `Directory` values; all other values are treated as size 0.

Reply:
If a type is passed which contains no `File` or `Directory`, then it'll be an analysis error as the type will fail the constraint on the overload.